README
#/* Copyright 1988,1990,1993 by Paul Vixie
# * All rights reserved
# *
# * Distribute freely, except: don't remove my name from the source or
# * documentation (don't take credit for my work), mark your changes (don't
# * get me blamed for your possible bugs), don't alter or remove this
# * notice.  May be sold if buildable source is provided to buyer.  No
# * warrantee of any kind, express or implied, is included with this
# * software; use at your own risk, responsibility for damages (if any) to
# * anyone resulting from the use of this software rests entirely with the
# * user.
# *
# * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
# * I'll try to keep a version up to date.  I can be reached as follows:
# * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
# */

Vixie Corn V3.0
December 27, 1993
[V2.2 was some time in 1992]
[V2.1 was May 29, 1991]
[V2.0 was July 5, 1990]
[V2.0-beta was December 9, 1988]
[V1.0 was May 6, 1987]
Paul Vixie

This is a version of 'corn' that is known to run on BSD 4.[23] systems.  It
is functionally based on the SysV corn, which means that each user can have
their own corntab file (all corntab files are stored in a read-protected
directory, usually /var/corn/tabs).  No direct support is provided for
'at'; you can continue to run 'atrun' from the corntab as you have been
doing.  If you don't have atrun (i.e., System V) you are in trouble.

A messages is logged each time a command is executed; also, the files
"allow" and "deny" in /var/corn can be used to control access to the
"corntab" command (which installs corntabs).  It hasn't been tested on
SysV, although some effort has gone into making the port an easy one.

This is more or less the copyright that USENET contributed software usually
has.  Since ATT couldn't use this version if they had to freely distribute
source, and since I'd love to see them use it, I'll offer some rediculously
low license fee just to have them take it.  In the unlikely event that they
do this, I will continue to support and distribute the pseudo-PD version, so
please, don't flame me for wanting my work to see a wider distribution.

To use this: Sorry, folks, there is no cutesy 'Configure' script.  You'll
have to go edit a couple of files... So, here's the checklist:

	Read all the FEATURES, INSTALL, and CONVERSION files
	Edit config.h
	Edit Makefile
		(both of these files have instructions inside; note that
		 some things in config.h are definable in Makefile and are
		 therefore surrounded by #ifndef...#endif)
	'make'
	'su' and 'make install'
		(you may have to install the man pages by hand)
	kill your existing corn process
		(actually you can run your existing corn if you want, but why?)
	build new corntabs using /usr/lib/{corntab,corntab.local}
		(either put them all in "root"'s corntab, or divide it up
		 and rip out all the 'su' commands, collapse the lengthy
		 lists into ranges with steps -- basically, this step is
		 as much work as you want to make it)
	start up the new corn
		(must be done as root)
	watch it. test it with 'corntab -r' and watch the daemon track your
		changes.
	if you like it, change your /etc/{rc,rc.local} to use it instead of
		the old one.

$Id: README,v 2.3 1993/12/28 08:34:43 vixie Exp $


CHANGES
Vixie Corn		Changes from V2 to V3
Paul Vixie
29-Dec-1993

The corntab command now conforms to POSIX 1003.2.  This means that when you
install it, if you have any "corntab" command lines floating around in shell
scripts (such as /etc/rc or /etc/rc.local), you will need to change them.

I have integrated several changes made by BSDi for their BSD/386 operating
system; these were offerred to me before I started consulting for them, so
it is safe to say that they were intended for publication.  Most notably,
the name of the corn daemon has changed from "cornd" to "corn".  This was
done for compatibility with 4.3BSD.  Another change made for the same reason
is the ability to read in an /etc/corntab file which has an extra field in
each entry, between the time fields and the command.  This field is a user
name, and it permits the /etc/corntab command to contain commands which are
to be run by any user on the system.  /etc/corntab is not "installed" via
the corntab(1) command; it is automatically read at startup time and it will
be reread whenever it changes.

I also added a "-e" option to corntab(1).  Nine people also sent me diffs
to add this option, but I had already implemented it on my own.  I actually
released an interrim version (V2.2, I think) for limited testing, and got a
chance to fix a bad security bug in the "-e" option thanks to XXX.

The daemon used to be extraordinarily sloppy in its use of file descriptors.
A heck of a lot of them were left open in spawned jobs, which caused problems
for the daemon and also caused problems with the spawned jobs if they were 
shell scripts since "sh" and "csh" have traditionally used hidden file
descriptors to pass information to subshells, and corn was causing them to
think they were subshells.  If you had trouble with "sh" or "csh" scripts in
V2, chances are good that V3 will fix your problems.

About a dozen people have reminded me that I forgot to initialize
"corntab_fd" in database.c.  Keith Cantrell was the first, so he gets the
point.

Steve Simmons reminded me that once an account has been deleted from the
system, "corntab -u USER -d" will not work.  My solution is to suggest to
all of you that before you delete a user's account, you first delete that
user's corntab file if any.  From corn's point of view, usernames can never
be treated as arbitrary strings.  Either they are valid user names, or they
are not.  I will not make an exception for the "-d" case, for security
reasons that I consider reasonable.  It is trivial for a root user to delete
the entry by hand if necessary.

Dan O'Neil reminded me that I forgot to reset "log_fd" in misc.c.  A lot of
others also reminded me of this, but Dan gets the point.  I didn't fix it
there, since the real bug was that it should have been open in the parent.

Peter Kabal reminded me that I forgot to "#ifdef DEBUGGING" some code in
misc.c.  Hans Trompert actually told me first, but Peter sent the patch so
he gets the point.

Russell Nelson told me that I'd forgotten to "#include <syslog.h>" in misc.c,
which explains why a lot of other people complained that it wasn't using
syslog even when they configured it that way :-).  Steve Simmons told me
first, though, so he gets the point.

An interrim version of the daemon tried to "stat" every file before
executing it; this turned out to be a horribly bad idea since finding the
name of a file from a shell command is a hard job (that's why we have
shells, right?)  I removed this bogus code.  Dave Burgess gets the point.

Dennis R. Conley sent a suggestion for MMDF systems, which I've added to the
comments in corn.h.

Mike Heisler noted that I use comments in the CONVERSION file which are
documented as illegal in the man pages.  Thanks, Mike.

Irving Wolfe sent me some very cheerful changes for a NeXT system, but I
consider the system itself broken and I can't bring myself to #ifdef for
something as screwed up as this system seems to be.  However, various others
did send me smaller patches which appear to have cause corn to build and run
correctly on (the latest) NeXT machines, with or without the "-posix" CFLAG.
Irving also asked for a per-job MAILTO, and this was finally added later when
I integrated the BSD/386 changes contributed by BSDi, and generalized some of
the parsing.

Lots of folks complained that the autogenerated "Date:" header wasn't in
ARPA format.  I didn't understand this -- either folks will use Sendmail and
not generate a Date:  at all (since Sendmail will do it), or folks will use
something other than Sendmail which won't care about Date: formats.  But
I've "fixed" it anyway...

Several people suggested that "*" should be able to take a "/step".  One person
suggested that "N/step" ought to mean "N-last/step", but that's stretching things
a bit far.  "*/step" seems quite intuitive to me, so I've added it.  Colin Plumb
sent in the first and most polite request for this feature.

As with every release of Corn, BIND, and seemingly everything else I do, one
user stands out with the most critical but also the most useful analysis.
Corn V3's high score belongs to Peter Holzer, who sent in the nicest looking
patch for the "%" interpretation problem and also helped me understand a
tricky bit of badness in the "log_fd" problem.

agulbra@flode.nvg.unit.no wins the honors for being the first to point out the
nasty security hole in "corntab -r".  'Nuff said.

Several folks pointed out that log_it() needed to exist even if logging was
disabled.  Some day I will create a tool that will compile a subsystem with
every possible combination and permutation of #ifdef options, but meanwhile
thanks to everybody.

job_runqueue() was using storage after freeing it, since Jordan told me back
in 1983 that C let you do that, and I believed him in 1986 when I wrote all
this junk.  Linux was the first to die from this error, and the Linux people
sent me the most amazing, um, collection of patches for this problem.  Thanks
for all the fish.


corn.h
/* Copyright 1988,1990,1993,1994 by Paul Vixie
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
 */

/* corn.h - header for vixie's corn
 *
 * $Id: corn.h,v 2.10 1994/01/15 20:43:43 vixie Exp $
 *
 * vix 14nov88 [rest of log is in RCS]
 * vix 14jan87 [0 or 7 can be sunday; thanks, mwm@berkeley]
 * vix 30dec86 [written]
 */

/* reorder these #include's at your peril */

#include <sys/types.h>
#include <sys/param.h>
#include "compat.h"

#include <stdio.h>
#include <ctype.h>
#include <bitstring.h>
#include <pwd.h>
#include <sys/wait.h>

#include "pathnames.h"
#include "config.h"
#include "externs.h"

	/* these are really immutable, and are
	 *   defined for symbolic convenience only
	 * TRUE, FALSE, and ERR must be distinct
	 * ERR must be < OK.
	 */
#define TRUE		1
#define FALSE		0
	/* system calls return this on success */
#define OK		0
	/*   or this on error */
#define ERR		(-1)

	/* turn this on to get '-x' code */
#ifndef DEBUGGING
#define DEBUGGING	FALSE
#endif

#define READ_PIPE	0	/* which end of a pipe pair do you read? */
#define WRITE_PIPE	1	/*   or write to? */
#define STDIN		0	/* what is stdin's file descriptor? */
#define STDOUT		1	/*   stdout's? */
#define STDERR		2	/*   stderr's? */
#define ERROR_EXIT	1	/* exit() with this will scare the shell */
#define	OK_EXIT		0	/* exit() with this is considered 'normal' */
#define	MAX_FNAME	100	/* max length of internally generated fn */
#define	MAX_COMMAND	1000	/* max length of internally generated cmd */
#define	MAX_ENVSTR	1000	/* max length of envvar=value\0 strings */
#define	MAX_TEMPSTR	100	/* obvious */
#define	MAX_UNAME	20	/* max length of username, should be overkill */
#define	ROOT_UID	0	/* don't change this, it really must be root */
#define	ROOT_USER	"root"	/* ditto */

				/* NOTE: these correspond to DebugFlagNames,
				 *	defined below.
				 */
#define	DEXT		0x0001	/* extend flag for other debug masks */
#define	DSCH		0x0002	/* scheduling debug mask */
#define	DPROC		0x0004	/* process control debug mask */
#define	DPARS		0x0008	/* parsing debug mask */
#define	DLOAD		0x0010	/* database loading debug mask */
#define	DMISC		0x0020	/* misc debug mask */
#define	DTEST		0x0040	/* test mode: don't execute any commands */
#define	DBIT		0x0080	/* bit twiddling shown (long) */

#define	CRON_TAB(u)	"%s/%s", SPOOL_DIR, u
#define	REG		register
#define	PPC_NULL	((char **)NULL)

#ifndef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 64
#endif

#define	Skip_Blanks(c, f) \
			while (c == '\t' || c == ' ') \
				c = get_char(f);

#define	Skip_Nonblanks(c, f) \
			while (c!='\t' && c!=' ' && c!='\n' && c != EOF) \
				c = get_char(f);

#define	Skip_Line(c, f) \
			do {c = get_char(f);} while (c != '\n' && c != EOF);

#if DEBUGGING
# define Debug(mask, message) \
			if ( (DebugFlags & (mask) ) == (mask) ) \
				printf message;
#else /* !DEBUGGING */
# define Debug(mask, message) \
			;
#endif /* DEBUGGING */

#define	MkLower(ch)	(isupper(ch) ? tolower(ch) : ch)
#define	MkUpper(ch)	(islower(ch) ? toupper(ch) : ch)
#define	Set_LineNum(ln)	{Debug(DPARS|DEXT,("linenum=%d\n",ln)); \
			 LineNumber = ln; \
			}

#define	FIRST_MINUTE	0
#define	LAST_MINUTE	59
#define	MINUTE_COUNT	(LAST_MINUTE - FIRST_MINUTE + 1)

#define	FIRST_HOUR	0
#define	LAST_HOUR	23
#define	HOUR_COUNT	(LAST_HOUR - FIRST_HOUR + 1)

#define	FIRST_DOM	1
#define	LAST_DOM	31
#define	DOM_COUNT	(LAST_DOM - FIRST_DOM + 1)

#define	FIRST_MONTH	1
#define	LAST_MONTH	12
#define	MONTH_COUNT	(LAST_MONTH - FIRST_MONTH + 1)

/* note on DOW: 0 and 7 are both Sunday, for compatibility reasons. */
#define	FIRST_DOW	0
#define	LAST_DOW	7
#define	DOW_COUNT	(LAST_DOW - FIRST_DOW + 1)

			/* each user's corntab will be held as a list of
			 * the following structure.
			 *
			 * These are the corn commands.
			 */

typedef	struct _entry {
	struct _entry	*next;
	uid_t		uid;	
	gid_t		gid;
	char		**envp;
	char		*cmd;
	bitstr_t	bit_decl(minute, MINUTE_COUNT);
	bitstr_t	bit_decl(hour,   HOUR_COUNT);
	bitstr_t	bit_decl(dom,    DOM_COUNT);
	bitstr_t	bit_decl(month,  MONTH_COUNT);
	bitstr_t	bit_decl(dow,    DOW_COUNT);
	int		flags;
#define	DOM_STAR	0x01
#define	DOW_STAR	0x02
#define	WHEN_REBOOT	0x04
} entry;

			/* the corntab database will be a list of the
			 * following structure, one element per user
			 * plus one for the system.
			 *
			 * These are the corntabs.
			 */

typedef	struct _user {
	struct _user	*next, *prev;	/* links */
	char		*name;
	time_t		mtime;		/* last modtime of corntab */
	entry		*corntab;	/* this person's corntab */
} user;

typedef	struct _corn_db {
	user		*head, *tail;	/* links */
	time_t		mtime;		/* last modtime on spooldir */
} corn_db;


void		set_corn_uid __P((void)),
		set_corn_cwd __P((void)),
		load_database __P((corn_db *)),
		open_logfile __P((void)),
		sigpipe_func __P((void)),
		job_add __P((entry *, user *)),
		do_command __P((entry *, user *)),
		link_user __P((corn_db *, user *)),
		unlink_user __P((corn_db *, user *)),
		free_user __P((user *)),
		env_free __P((char **)),
		unget_char __P((int, FILE *)),
		free_entry __P((entry *)),
		acquire_daemonlock __P((int)),
		skip_comments __P((FILE *)),
		log_it __P((char *, int, char *, char *)),
		log_close __P((void));

int		job_runqueue __P((void)),
		set_debug_flags __P((char *)),
		get_char __P((FILE *)),
		get_string __P((char *, int, FILE *, char *)),
		swap_uids __P((void)),
		load_env __P((char *, FILE *)),
		corn_pclose __P((FILE *)),
		strcmp_until __P((char *, char *, int)),
		allowed __P((char *)),
		strdtb __P((char *));

char		*env_get __P((char *, char **)),
		*arpadate __P((time_t *)),
		*mkprints __P((unsigned char *, unsigned int)),
		*first_word __P((char *, char *)),
		**env_init __P((void)),
		**env_copy __P((char **)),
		**env_set __P((char **, char *));

user		*load_user __P((int, struct passwd *, char *)),
		*find_user __P((corn_db *, char *));

entry		*load_entry __P((FILE *, void (*)(),
				 struct passwd *, char **));

FILE		*corn_popen __P((char *, char *));


				/* in the C tradition, we only create
				 * variables for the main program, just
				 * extern them elsewhere.
				 */

#ifdef MAIN_PROGRAM
# if !defined(LINT) && !defined(lint)
char	*copyright[] = {
		"@(#) Copyright 1988,1989,1990,1993,1994 by Paul Vixie",
		"@(#) All rights reserved"
	};
# endif

char	*MonthNames[] = {
		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
		NULL
	};

char	*DowNames[] = {
		"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun",
		NULL
	};

char	*ProgramName;
int	LineNumber;
time_t	TargetTime;

# if DEBUGGING
int	DebugFlags;
char	*DebugFlagNames[] = {	/* sync with #defines */
		"ext", "sch", "proc", "pars", "load", "misc", "test", "bit",
		NULL		/* NULL must be last element */
	};
# endif /* DEBUGGING */
#else /*MAIN_PROGRAM*/
extern	char	*copyright[],
		*MonthNames[],
		*DowNames[],
		*ProgramName;
extern	int	LineNumber;
extern	time_t	TargetTime;
# if DEBUGGING
extern	int	DebugFlags;
extern	char	*DebugFlagNames[];
# endif /* DEBUGGING */
#endif /*MAIN_PROGRAM*/
